File: threads/src/main/java/im/threads/internal/database/DatabaseHolder.java
package im.threads.internal.database;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.List;

import im.threads.internal.Config;
import im.threads.internal.model.ChatItem;
import im.threads.internal.model.ConsultInfo;
import im.threads.internal.model.ConsultPhrase;
import im.threads.internal.model.FileDescription;
import im.threads.internal.model.MessageState;
import im.threads.internal.model.Survey;
import im.threads.internal.model.UserPhrase;
import io.reactivex.Completable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

public final class DatabaseHolder {

    private static DatabaseHolder instance;
    private final ThreadsDbHelper mMyOpenHelper;

    private DatabaseHolder() {
        mMyOpenHelper = new ThreadsDbHelper(Config.instance.context);
    }

    @NonNull
    public static DatabaseHolder getInstance() {
        if (instance == null) {
            instance = new DatabaseHolder();
        }
        return instance;
    }

    // ChatItems

    /**
     * Nullify instance. For Autotests purposes
     */
    static void eraseInstance() {
        instance = null;
    }

    public void cleanDatabase() {
        mMyOpenHelper.cleanDatabase();
    }

    public List<ChatItem> getChatItems(int offset, int limit) {
        return mMyOpenHelper.getChatItems(offset, limit);
    }

    @Nullable
    public ChatItem getChatItem(String messageUuid) {
        return mMyOpenHelper.getChatItem(messageUuid);
    }

    public void putChatItems(final List<ChatItem> items) {
        mMyOpenHelper.putChatItems(items);
    }

    public boolean putChatItem(ChatItem chatItem) {
        return mMyOpenHelper.putChatItem(chatItem);
    }

    // FileDescriptions

    public Single<List<FileDescription>> getAllFileDescriptions() {
        return Single.fromCallable(mMyOpenHelper::getAllFileDescriptions)
                .subscribeOn(Schedulers.io());
    }

    // UserPhrase

    public void updateFileDescription(@NonNull FileDescription fileDescription) {
        mMyOpenHelper.updateFileDescription(fileDescription);
    }

    @Nullable
    public ConsultInfo getConsultInfo(@NonNull String id) {
        return mMyOpenHelper.getLastConsultInfo(id);
    }

    public List<UserPhrase> getUnsendUserPhrase(int count) {
        return mMyOpenHelper.getUnsendUserPhrase(count);
    }

    // ConsultPhrase

    public void setStateOfUserPhraseByProviderId(String providerId, MessageState messageState) {
        mMyOpenHelper.setUserPhraseStateByProviderId(providerId, messageState);
    }

    public Single<ConsultPhrase> getLastConsultPhrase() {
        return Single.fromCallable(mMyOpenHelper::getLastConsultPhrase)
                .subscribeOn(Schedulers.io());
    }

    public Completable setAllConsultMessagesWereRead() {
        return Completable.fromCallable(mMyOpenHelper::setAllConsultMessagesWereRead)
                .subscribeOn(Schedulers.io());
    }

    public void setConsultMessageWasRead(String providerId) {
        mMyOpenHelper.setConsultMessageWasRead(providerId);
    }

    // Survey
    @Nullable
    public Survey getSurvey(long sendingId) {
        return mMyOpenHelper.getSurvey(sendingId);
    }

    public Completable setOldSurveyDisplayMessageToFalse() {
        return Completable.fromCallable(mMyOpenHelper::setOldSurveyDisplayMessageToFalse)
                .subscribeOn(Schedulers.io());
    }

    // Messages
    public int getMessagesCount() {
        return mMyOpenHelper.getMessagesCount();
    }

    // let the DB time to write the incoming message
    public int getUnreadMessagesCount() {
        return mMyOpenHelper.getUnreadMessagesCount();
    }

    @NonNull
    public List<String> getUnreadMessagesUuid() {
        return mMyOpenHelper.getUnreadMessagesUuid();
    }

    /**
     * For Autotests purposes
     *
     * @return MyOpenHelper instance
     */
    ThreadsDbHelper getMyOpenHelper() {
        return mMyOpenHelper;
    }


}

File: src/main/java/im/threads/business/secureDatabase/DatabaseHolder.kt
package im.threads.business.secureDatabase

import android.content.Context
import android.database.sqlite.SQLiteDiskIOException
import im.threads.R
import im.threads.business.annotation.OpenForTesting
import im.threads.business.models.ChatItem
import im.threads.business.models.ConsultInfo
import im.threads.business.models.ConsultPhrase
import im.threads.business.models.FileDescription
import im.threads.business.models.MessageStatus
import im.threads.business.models.SpeechMessageUpdate
import im.threads.business.models.Survey
import im.threads.business.models.UserPhrase
import im.threads.business.utils.Balloon
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers

@OpenForTesting
class DatabaseHolder(private val context: Context) {

    init {
        checkAndUpdate()
    }

    private var myOpenHelper = ThreadsDbHelper.getInstance(context)

    fun cleanDatabase() = tryExecute { myOpenHelper.cleanDatabase() }

    fun getChatItems(offset: Int, limit: Int): List<ChatItem> {
        return tryExecute {
            val items = myOpenHelper.getChatItems(offset, limit).toMutableList()
            val surveysWithQuestions = mutableListOf<Survey>()
            items.filterIsInstance<Survey>().forEach { survey ->
                (myOpenHelper.getChatItemByCorrelationId(survey.uuid) as? Survey)?.let { surveyWithQuestions ->
                    surveysWithQuestions.add(surveyWithQuestions)
                }
            }
            surveysWithQuestions.forEach { surveyWithQuestions ->
                val indexOfSurvey = items.indexOfFirst { it is Survey && it.sendingId == surveyWithQuestions.sendingId }
                if (indexOfSurvey >= 0) {
                    items[indexOfSurvey] = surveyWithQuestions
                }
            }
            items
        } ?: arrayListOf()
    }

    fun getSendingChatItems(): List<UserPhrase> = tryExecute { myOpenHelper.getSendingChatItems() } ?: arrayListOf()

    fun getNotDeliveredChatItems(): List<UserPhrase> = tryExecute { myOpenHelper.getNotDeliveredChatItems() } ?: arrayListOf()

    fun getChatItemByCorrelationId(messageUuid: String?): ChatItem? =
        tryExecute { myOpenHelper.getChatItemByCorrelationId(messageUuid) }

    fun getChatItemByBackendMessageId(messageId: String?): ChatItem? =
        tryExecute { myOpenHelper.getChatItemByBackendMessageId(messageId) }

    fun putChatItems(items: List<ChatItem?>?) = tryExecute { myOpenHelper.putChatItems(items) }

    fun putChatItem(chatItem: ChatItem?): Boolean = tryExecute { myOpenHelper.putChatItem(chatItem) } ?: false

    // FileDescriptions
    val allFileDescriptions: Single<List<FileDescription?>?>
        get() = Single.fromCallable { tryExecute { myOpenHelper.getAllFileDescriptions() } }
            .subscribeOn(Schedulers.io())

    // UserPhrase
    fun updateFileDescription(fileDescription: FileDescription) {
        tryExecute { myOpenHelper.updateFileDescription(fileDescription) }
    }

    fun updateChatItemByTimeStamp(chatItem: ChatItem) {
        tryExecute { myOpenHelper.updateChatItemByTimeStamp(chatItem) }
    }

    fun getConsultInfo(id: String): ConsultInfo? = tryExecute { myOpenHelper.getLastConsultInfo(id) }

    fun getUnsendUserPhrase(count: Int): List<UserPhrase> = tryExecute { myOpenHelper.getUnsendUserPhrase(count) } ?: arrayListOf()

    fun setStateOfUserPhraseByCorrelationId(uuid: String?, messageStatus: MessageStatus?) {
        tryExecute { myOpenHelper.setUserPhraseStateByCorrelationId(uuid, messageStatus) }
    }

    fun setStateOfUserPhraseByBackendMessageId(messageId: String?, messageStatus: MessageStatus?) {
        tryExecute { myOpenHelper.setUserPhraseStateByBackendMessageId(messageId, messageStatus) }
    }

    val lastConsultPhrase: Single<ConsultPhrase?> =
        Single.fromCallable { tryExecute { myOpenHelper.getLastConsultPhrase() } }
            .subscribeOn(Schedulers.io())

    fun setAllConsultMessagesWereRead(): Completable {
        return Completable.fromCallable { tryExecute { myOpenHelper.setAllConsultMessagesWereRead() } }
            .subscribeOn(Schedulers.io())
    }

    fun setAllConsultMessagesWereReadInThread(threadId: Long?): Completable {
        return Completable.fromCallable { tryExecute { myOpenHelper.setAllConsultMessagesWereReadWithThreadId(threadId) } }
            .subscribeOn(Schedulers.io())
    }

    fun setMessageWasRead(uuid: String?) {
        uuid?.let { tryExecute { myOpenHelper.setMessageWasRead(it) } }
    }

    fun saveSpeechMessageUpdate(speechMessageUpdate: SpeechMessageUpdate?) {
        tryExecute { myOpenHelper.speechMessageUpdated(speechMessageUpdate) }
    }

    fun setNotSentSurveyDisplayMessageToFalse(): Completable {
        return Completable.fromCallable { tryExecute { myOpenHelper.setNotSentSurveyDisplayMessageToFalse() } }
            .subscribeOn(Schedulers.io())
    }

    fun setOldRequestResolveThreadDisplayMessageToFalse(): Completable {
        return Completable.fromCallable { tryExecute { myOpenHelper.setOldRequestResolveThreadDisplayMessageToFalse() } }
            .subscribeOn(Schedulers.io())
    }

    fun getMessagesCount(): Int = tryExecute { myOpenHelper.getMessagesCount() } ?: 0

    fun getUnreadMessagesCount(): Int = tryExecute { myOpenHelper.getUnreadMessagesCount() } ?: 0

    fun getUnreadMessagesUuid(): List<String?> = tryExecute { myOpenHelper.getUnreadMessagesUuid() } ?: arrayListOf()

    fun setOrUpdateMessageId(correlationId: String?, backendMessageId: String?) {
        tryExecute { myOpenHelper.setOrUpdateMessageId(correlationId, backendMessageId) }
    }

    fun removeItem(correlationId: String?, messageId: String?) = tryExecute {
        myOpenHelper.removeItem(correlationId, messageId)
    }

    private fun checkIsDatabaseCorrupted() {
        myOpenHelper = ThreadsDbHelper.getInstance(context)
    }

    private fun checkAndUpdate() {
        val oldHelper = im.threads.business.database.ThreadsDbHelper(context)
        if (needMigrateToNewDB(oldHelper)) {
            putChatItems(oldHelper.getChatItems(0, -1))
            myOpenHelper.putFileDescriptions(oldHelper.allFileDescriptions)
            oldHelper.cleanDatabase()
        }
    }

    private fun needMigrateToNewDB(helper: im.threads.business.database.ThreadsDbHelper): Boolean {
        return try {
            helper.getChatItems(0, -1).size > 0 || helper.allFileDescriptions.size > 0
        } catch (exc: SQLiteDiskIOException) {
            Balloon.show(context, context.getString(R.string.ecc_not_enough_space))
            false
        }
    }

    private fun <T> tryExecute(block: () -> T?): T? {
        return try {
            block()
        } catch (exc: Exception) {
            checkIsDatabaseCorrupted()
            block()
        }
    }
}